README.md
# Money

[![Build Status][build status badge]][build status]
[![License][license badge]][license]
[![Swift Version][swift version badge]][swift version]
![Cocoapods platforms][cocoapods platforms badge]
[![Cocoapods compatible][cocoapods badge]][cocoapods]
[![Carthage compatible][carthage badge]][carthage]

A precise, type-safe representation of monetary amounts in a given currency.

This functionality is discussed in Chapter 3 of
[Flight School Guide to Swift Numbers](https://flight.school/books/numbers).

## Requirements

- Swift 4.0+

## Installation

### Swift Package Manager

Add the Money package to your target dependencies in `Package.swift`:

```swift
import PackageDescription

let package = Package(
  name: "YourProject",
  dependencies: [
    .package(
        url: "https://github.com/Flight-School/Money",
        from: "1.4.0"
    ),
  ]
)
```

Then run the `swift build` command to build your project.

### CocoaPods

You can install `Money` via CocoaPods,
by adding the following line to your `Podfile`:

```ruby
pod 'Money-FlightSchool', '~> 1.4.0'
```

Run the `pod install` command to download the library
and integrate it into your Xcode project.

> **Note**
> The module name for this library is "Money" ---
> that is, to use it, you add `import Money` to the top of your Swift code
> just as you would by any other installation method.
> The pod is called "Money-FlightSchool"
> because there's an existing pod with the name "Money".

### Carthage

To use `Money` in your Xcode project using Carthage,
specify it in `Cartfile`:

```
github "Flight-School/Money" ~> 1.4.0
```

Then run the `carthage update` command to build the framework,
and drag the built Money.framework into your Xcode project.

## Usage

### Creating Monetary Amounts

The `Money` type has a required associated `Currency` type.
These currency types are named according to their
three letter [ISO 4701][iso4217] currency codes.
You can initialize a monetary using a `Decimal` value:

```swift
let amount = Decimal(12)
let monetaryAmount = Money<USD>(amount)
```

Some currencies specify a minor unit.
For example, USD amounts are often expressed in cents,
each worth 1/100 of a dollar.
You can initialize monetary amounts from a quantity of minor units.
For currencies that don't have a minor unit,
such as JPY,
this is equivalent to the standard initializer.

```swift
let twoCents = Money<USD>(minorUnits: 2)
twoCents.amount // 0.02

let ichimonEn = Money<JPY>(minorUnits: 10_000)
ichimonEn.amount // 10000
```

You can also create monetary amounts using
integer, floating-point, and string literals.

```swift
12 as Money<USD>
12.00 as Money<USD>
"12.00" as Money<USD>
```

**Important**:
Swift floating-point literals are currently initialized
using binary floating-point number type,
which cannot precisely express certain values.
As a workaround, monetary amounts initialized
from a floating-point literal are rounded
to the number of places of the minor currency unit.
If you want to express a smaller fractional monetary amount,
initialize from a string literal or `Decimal` value instead.

```swift
let preciseAmount: Money<USD> = "123.4567"
let roundedAmount: Money<USD> = 123.4567

preciseAmount.amount // 123.4567
roundedAmount.amount // 123.46
```

For more information, see https://bugs.swift.org/browse/SR-920.

### Comparing Monetary Amounts

You can compare two monetary amounts with the same currency:

```swift
let amountInWallet: Money<USD> = 60.00
let price: Money<USD> = 19.99

amountInWallet >= price // true
```

Attempting to compare monetary amounts with different currencies
results in a compiler error:

```swift
let dollarAmount: Money<USD> = 123.45
let euroAmount: Money<EUR> = 4567.89

dollarAmount == euroAmount // Error: Binary operator '==' cannot be applied
```

### Adding, Subtracting, and Multiplying Monetary Amounts

Monetary amounts can be added, subtracted, and multiplied
using the standard binary arithmetic operators (`+`, `-`, `*`):

```swift
let prices: [Money<USD>] = [2.19, 5.39, 20.99, 2.99, 1.99, 1.99, 0.99]
let subtotal = prices.reduce(0.00, +) // "$36.53"
let tax = 0.08 * subtotal // "$2.92"
let total = subtotal + tax // "$39.45"
```

**Important**: Multiplying a monetary amount by a floating-point number
results in an amount rounded to the number of places
of the minor currency unit.
If you want to produce a smaller fractional monetary amount,
multiply by a `Decimal` value instead.

### Formatting Monetary Amounts

You can create a localized representation of a monetary amount
using `NumberFormatter`.
Set the `currencyCode` property of the formatter
to the `currency.code` property of the `Money` value
and pass the `amount` property to the formatter `string(for:)` method.

```swift
let allowance: Money<USD> = 10.00
let formatter = NumberFormatter()
formatter.numberStyle = .currency
formatter.locale = Locale(identifier: "fr-FR")
formatter.currencyCode = allowance.currency.code
formatter.string(for: allowance.amount) // "10,00 $US"
```

### Encoding and Decoding Monetary Amounts

#### Encoding

By default,
`Money` values are encoded as keyed containers,
with `amount` encoded as a number value.

```swift
let value: Money<USD> = 123.45

let encoder = JSONEncoder()
let data = try encoder.encode(value)
String(data: data, encoding: .utf8) // #"{"amount":123.45,"currencyCode":"USD"}"#
```

To configure encoding behavior,
set either the `JSONEncoder.moneyEncodingOptions` property
or the `CodingUserInfoKey.moneyEncodingOptions` key
in the encoder's `userInfo` property.

```swift
var encoder = JSONEncoder()
encoder.moneyEncodingOptions = [.omitCurrency, .encodeAmountAsString]

let data = try encoder.encode([value])
String(data: data, encoding: .utf8) // #"["123.45"]"#
```

#### Decoding

The default decoding behavior is flexible,
supporting both keyed and single value containers,
with string or number values for `amount`.

```swift
let json = #"""
[
    { "currencyCode": "USD", "amount": "100.00" },
    50.00,
    "10"
]
"""#.data(using: .utf8)!

let decoder = JSONDecoder()
let values = try decoder.decode([Money<USD>].self, from: json)
values.first?.amount // 100.00
values.last?.currency.code // "USD"
```

To configure decoding behavior,
set either the `JSONDecoder.moneyDecodingOptions` property
or the `CodingUserInfoKey.moneyDecodingOptions` key
in the decoder's `userInfo` property.

```swift
var decoder = JSONDecoder()
decoder.moneyDecodingOptions = [.requireExplicitCurrency]
```

**Important**:
Foundation decoders currently decode number values
using a binary floating-point number type,
which cannot precisely express certain values.
As a workaround,
you can specify the `requireStringAmount` decoding option
to require monetary amounts to be
decoded precisely from a string representation.

```swift
let json = #"""
{ "currencyCode": "USD", "amount": "27.31" }
"""#.data(using: .utf8)!

var decoder = JSONDecoder()

try decoder.decode(Money<USD>.self, from: json) // DecodingError

decoder.moneyDecodingOptions = [.requireStringAmount]
let preciseAmount = try decoder.decode(Money<USD>.self, from: json)
preciseAmount.amount // 27.31
```

Alternatively,
you can the `roundFloatingPointAmount` decoding option
to round decoded floating-point values
to the number of places of the minor currency unit.

```swift
let json = #"""
{ "currencyCode": "USD", "amount": 27.31 }
"""#.data(using: .utf8)!

var decoder = JSONDecoder()

let impreciseAmount = try decoder.decode(Money<USD>.self, from: json)
impreciseAmount.amount // 27.30999999...

decoder.moneyDecodingOptions = [.roundFloatingPointAmount]
let roundedAmount = try decoder.decode(Money<USD>.self, from: json)
roundedAmount.amount // 27.31
```

For more information, see https://bugs.swift.org/browse/SR-7054.

#### Customizing Coding Keys

By default,
`Money` values are encoded and decoded with the string keys
`"amount"` and `"currencyCode"`, which
correspond to their respective properties.

If you're working with data that encodes monetary amounts differently,
you can set the `keyDecodingStrategy` property of `JSONDecoder`
to map to different key names:

```swift
let json = #"""
 {
    "value": "3.33",
    "currency": "USD"
 }
 """#.data(using: .utf8)!

let decoder = JSONDecoder()
decoder.keyDecodingStrategy = .custom({ keys in
    switch keys.last?.stringValue {
    case "value":
        return MoneyCodingKeys.amount
    case "currency":
        return MoneyCodingKeys.currencyCode
    default:
        return keys.last!
    }
})

let amount = try decoder.decode(Money<USD>.self, from: json) // $3.33
```

Alternatively,
you can create structures that match the shape of your data
and derive computed properties that return `Money` types:

```swift
struct Item: Codable {
    struct Price: Codable {
        let value: String
        let currency: String
    }

    let name: String
    private let unitPrice: Price

    var unitPriceInUSD: Money<USD>? {
        guard unitPrice.currency == USD.code else { return nil }
        return Money(unitPrice.value)
    }
}

let json = #"""
 {
    "name": "Widget",
    "unitPrice": {
       "value": "3.33",
       "currency": "USD"
    }
 }
 """#.data(using: .utf8)!

let decoder = JSONDecoder()
let item = try decoder.decode(Item.self, from: json)
item.unitPriceInUSD // $3.33
```

### Supporting Multiple Currencies

Consider a `Product` structure with a `price` property.
If you only support a single currency, such as US Dollars,
you would define `price` to be of type `Money<USD>`:

```swift
struct Product {
    var price: Money<USD>
}
```

If you want to support multiple currencies, however,
you can't specify an explicit currency type in the property declaration.
Instead, the `Product` would have to be defined as a generic type:

```swift
struct Product<Currency: CurrencyType> {
    var price: Money<Currency>
}
```

Unfortunately, this approach is unwieldy,
as each type that interacts with `Product` would also need to be generic,
and so on, until the entire code base is generic over the currency type.

```swift
class ViewController<Currency: CurrencyType> : UIViewController { ... } // 😭
```

A better solution would be to define a new `Price` protocol
with requirements that match the `Money` type:

```swift
protocol Price {
    var amount: Decimal { get }
    var currency: CurrencyType.Type { get }
}

extension Money: Price {}
```

Doing this allows prices to be defined in multiple currencies
without making `Product` generic over the currency type:

```swift
struct Product {
    var price: Price
}

let product = Product(price: 12.00 as Money<USD>)
product.price // "$12.00"
```

If you want to support only certain currencies, such as US Dollars and Euros,
you can define a `SupportedCurrency` protocol
and add conformance to each currency type through an extension:

```swift
protocol SupportedCurrency: CurrencyType {}
extension USD: SupportedCurrency {}
extension EUR: SupportedCurrency {}

extension Money: Price where Currency: SupportedCurrency {}
```

Now, attempting to create a `Product` with a price in an unsupported currency
results in a compiler error:

```swift
Product(price: 100.00 as Money<EUR>)
Product(price: 100.00 as Money<GBP>) // Error
```

#### Supported Currencies

This package provides a `Currency` type for
each of the currencies defined by the [ISO 4217][iso4217] standard
with the exception of special codes,
such as USN (US Dollar, Next day) and
XBC (Bond Markets Unit European Unit of Account 9).

The [source file][currency.swift] defining the available currencies
is generated from a [CSV file][iso4217.csv] using [GYB][gyb].
This data source is up-to-date with
[ISO 4217 Amendment Number 169](https://www.currency-iso.org/en/shared/amendments/iso-4217-amendment.html),
published on August 17, 2018.

You can regenerate `Sources/Money/Currency.swift` from `Resources/iso4217.csv`
by installing [GYB][gyb]
and running the `make` command from the terminal:

```terminal
$ make
```

> We don't currently have a mechanism to automatically update this data source.
> Please [open an issue](https://github.com/Flight-School/Money/issues/new)
> if you're aware of any new amendments made to ISO 4217.

You can lookup any built-in currency types by its three-letter code
using the `iso4217Currency(for:)` function.

```swift
iso4217Currency(for: "USD")?.name // "US Dollar"
iso4217Currency(for: "invalid") // nil
```

### Adding Custom Currencies

You can create your own custom currency types by defining an enumeration
that conforms to the `CurrencyType` protocol.
For example, here's how you might represent Bitcoin (BTC):

```swift
enum BTC: CurrencyType {
    static var name: String { return "Bitcoin" }
    static var code: String { return "BTC" }
    static var minorUnit: Int { return 8 }
}

let satoshi: Money<BTC> = 0.00000001
```

`NumberFormatter` only supports currencies defined by ISO 4217,
so you'll have to configure the symbol, currency code,
and any other necessary parameters:

```swift
let formatter = NumberFormatter()
formatter.numberStyle = .currency
formatter.currencySymbol = "₿"
formatter.currencyCode = "BTC"
formatter.maximumFractionDigits = 8

formatter.string(for: satoshi.amount) // ₿0.00000001
```

> **Important**:
> The `iso4217Currency(for:) returns only built-in currencies,
> so calling `iso4217Currency(for: "BTC")` would return `nil`.

### Showing Off with Emoji

If you're the type of person who enjoys putting clip art in your source code,
here's a trick that'll _really_ impress your teammates:

```swift
typealias 💵 = Money<USD>
typealias 💴 = Money<JPY>
typealias 💶 = Money<EUR>
typealias 💷 = Money<GBP>

let tubeFare: 💷 = 2.40 // "£2.40"
```

## Alternatives to Consider

A type-safe `Money` structure like the one provided by this package
can reduce the likelihood of certain kinds of programming errors.
However, you may find the cost of using this abstraction
to outweigh the benefits it can provide in your code base.

If that's the case,
you might consider implementing your own simple `Money` type
with a nested `Currency` enumeration like this:

```swift
struct Money {
   enum Currency: String {
      case USD, EUR, GBP, CNY // supported currencies here
   }

   var amount: Decimal
   var currency: Currency
}
```

It's ultimately up to you to decide what kind of abstraction
is best for your particular use case.
Whatever you choose,
just make sure to represent monetary amounts using a `Decimal` type
with an explicit currency.

## License

MIT

## Contact

Mattt ([@mattt](https://twitter.com/mattt))

[build status]: https://github.com/Flight-School/Money/actions?query=workflow%3ACI
[build status badge]: https://github.com/Flight-School/Money/workflows/CI/badge.svg
[currency.swift]: https://github.com/Flight-School/Money/blob/main/Sources/Money/Currency.swift
[iso4217]: https://en.wikipedia.org/wiki/ISO_4217
[iso4217.csv]: https://github.com/Flight-School/Money/blob/main/Resources/iso4217.csv
[gyb]: https://nshipster.com/swift-gyb/
[license]: https://opensource.org/licenses/MIT
[license badge]: https://img.shields.io/cocoapods/l/Money-FlightSchool.svg
[swift version]: https://swift.org/download/
[swift version badge]: https://img.shields.io/badge/swift%20version-4.0+-orange.svg
[cocoapods platforms badge]: https://img.shields.io/cocoapods/p/Money-FlightSchool.svg
[cocoapods]: https://cocoapods.org/pods/Money-FlightSchool
[cocoapods badge]: https://img.shields.io/cocoapods/v/Money-FlightSchool.svg
[carthage]: https://github.com/Carthage/Carthage
[carthage badge]: https://img.shields.io/badge/Carthage-compatible-4BC51D.svg


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.4.0] - 2022-10-10

- Added `MoneyCodingKeys` type for more convenient decoding of custom keys.
  #21 by @mattt.

## [1.3.0] - 2021-04-17

### Added

- Added initializer from a currency's minor units.
  #15 by @mattt.
- Added top-level `iso4217Currency(for:)` function.
  #16 by @mattt.

## [1.2.1] - 2020-11-09

### Changed

- Changed Swift versions specified in CocoaPods podspec file.
  #11 by @mattt.

## [1.2.0] - 2019-05-06

### Added

- Added configurable encoding and decoding behavior.
  #9 by @mattt.

## [1.1.1] - 2019-05-11

### Added

- Added support for CocoaPods.
  d51b2f14 by @mattt.

## [1.1.0] - 2019-03-31

### Added

- Added `AdditiveArithmetic` protocol conformance in Swift 5.
  d1bbaf by @mattt.
- Added `CHANGELOG.md`.
  1caec0 by @mattt.
  
### Changed

- Changed Xcode project settings for Xcode 10.2 and Swift 5.
  facb5a4 by @mattt.

## [1.0.2] - 2019-03-31

### Added

- Added `Makefile` to automatically generate `Sources/Money/Currency.swift`.
  3cc0e87 by @mattt.
- Added `Brewfile` to install `gyb` command line tool.
  153cdb by @mattt.

### Changed

- Changed entry for Comorian Franc to remove extra space.
  1ba55a by @mattt.

### Removed

- Removed GYB tools from Vendor directory.
  938db3 by @mattt.

## [1.0.1] - 2018-07-03

### Added

- Added support for installation with Carthage.
  ed42a0 by @mattt.

## [1.0.0] - 2018-06-26

- Initial release

[unreleased]: https://github.com/SwiftDocOrg/doctest/compare/1.4.0...main
[1.3.0]: https://github.com/SwiftDocOrg/swift-doc/releases/tag/1.4.0
[1.3.0]: https://github.com/SwiftDocOrg/swift-doc/releases/tag/1.3.0
[1.2.1]: https://github.com/SwiftDocOrg/swift-doc/releases/tag/1.2.1
[1.2.0]: https://github.com/SwiftDocOrg/swift-doc/releases/tag/1.2.0
[1.1.1]: https://github.com/SwiftDocOrg/swift-doc/releases/tag/1.1.1
[1.1.0]: https://github.com/SwiftDocOrg/swift-doc/releases/tag/1.1.0
[1.0.2]: https://github.com/SwiftDocOrg/swift-doc/releases/tag/1.0.2
[1.0.1]: https://github.com/SwiftDocOrg/swift-doc/releases/tag/1.0.1
[1.0.0]: https://github.com/SwiftDocOrg/swift-doc/releases/tag/1.0.0


Tests/LinuxMain.swift
import XCTest

import MoneyTests

var tests = [XCTestCaseEntry]()
tests += MoneyTests.allTests()
XCTMain(tests)

Tests/MoneyTests/MoneyTests.swift
import XCTest
import Foundation
@testable import Money

final class MoneyTests: XCTestCase {
    func testMonetaryCalculations() {
        let prices: [Money<USD>] = [2.19, 5.39, 20.99, 2.99, 1.99, 1.99, 0.99]
        let subtotal = prices.reduce(0.00, +)
        let tax = 0.08 * subtotal
        let total = subtotal + tax
        
        XCTAssertEqual(subtotal.amount, Decimal(string: "36.53", locale: nil))
        XCTAssertEqual(tax.amount, Decimal(string: "2.92", locale: nil))
        XCTAssertEqual(total.amount, Decimal(string: "39.45", locale: nil))
    }

    func testMinorUnits() {
        let twoCents = Money<USD>(minorUnits: 2) // $0.02
        XCTAssertEqual(twoCents.amount, Decimal(string: "0.02", locale: nil))

        let ichimonEn = Money<JPY>(minorUnits: 10_000) // ¥10,000
        XCTAssertEqual(ichimonEn.amount, Decimal(string: "10000", locale: nil))
    }

    static var allTests = [
        ("testMonetaryCalculations", testMonetaryCalculations),
        ("testMinorUnits", testMinorUnits),
    ]
}


Tests/MoneyTests/CurrencyTests.swift
import XCTest
import Foundation
@testable import Money

final class CurrencyTests: XCTestCase {
    func testCurrencyByISO4217Code() {
        let usd = iso4217Currency(for: "USD")
        XCTAssert(usd == USD.self)
        XCTAssertEqual(usd?.code, "USD")
        XCTAssertEqual(usd?.name, "US Dollar")
        XCTAssertEqual(usd?.minorUnit, 2)

        let unk = iso4217Currency(for: "UNK")
        XCTAssertNil(unk)
    }

    static var allTests = [
        ("testCurrencyByISO4217Code", testCurrencyByISO4217Code),
    ]
}


Package.swift
// swift-tools-version:4.0
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "Money",
    products: [
        // Products define the executables and libraries produced by a package, and make them visible to other packages.
        .library(
            name: "Money",
            targets: ["Money"]),
    ],
    dependencies: [
        // Dependencies declare other packages that this package depends on.
        // .package(url: /* package url */, from: "1.0.0"),
    ],
    targets: [
        // Targets are the basic building blocks of a package. A target can define a module or a test suite.
        // Targets can depend on other targets in this package, and on products in packages which this package depends on.
        .target(
            name: "Money",
            dependencies: []),
        .testTarget(
            name: "MoneyTests",
            dependencies: ["Money"]),
    ]
)


Sources/Money/Money.swift
import Foundation

/// An amount of money in a given currency.
public struct Money<Currency: CurrencyType>: Equatable, Hashable {
    /// The amount of money.
    public var amount: Decimal

    /// Creates an amount of money with a given decimal number.
    public init(_ amount: Decimal) {
        self.amount = amount
    }

    public init(minorUnits: Int) {
        precondition(Currency.minorUnit >= 0)

        let sign: FloatingPointSign = minorUnits >= 0 ? .plus : .minus
        let exponent = -Currency.minorUnit
        let significand = Decimal(minorUnits)

        let amount = Decimal(sign: sign, exponent: exponent, significand: significand)
        self.init(amount)
    }

    /// The currency type.
    public var currency: CurrencyType.Type {
        return Currency.self
    }
    
    /**
        A monetary amount rounded to
        the number of places of the minor currency unit.
     */
    public var rounded: Money<Currency> {
        return Money<Currency>(amount.rounded(for: Currency.self))
    }
}

// MARK: - Comparable

extension Money: Comparable {
    public static func < (lhs: Money<Currency>, rhs: Money<Currency>) -> Bool {
        return lhs.amount < rhs.amount
    }
}

// MARK: - AdditiveArithmetic

#if swift(>=5.0)
extension Money: AdditiveArithmetic {}
#endif

extension Money {
    /// The sum of two monetary amounts.
    public static func + (lhs: Money<Currency>, rhs: Money<Currency>) -> Money<Currency> {
        return Money<Currency>(lhs.amount + rhs.amount)
    }

    /// Adds one monetary amount to another.
    public static func += (lhs: inout Money<Currency>, rhs: Money<Currency>) {
        lhs.amount += rhs.amount
    }

    /// The difference between two monetary amounts.
    public static func - (lhs: Money<Currency>, rhs: Money<Currency>) -> Money<Currency> {
        return Money<Currency>(lhs.amount - rhs.amount)
    }

    /// Subtracts one monetary amount from another.
    public static func -= (lhs: inout Money<Currency>, rhs: Money<Currency>) {
        lhs.amount -= rhs.amount
    }
}

extension Money {
    /// Negates the monetary amount.
    public static prefix func - (value: Money<Currency>) -> Money<Currency> {
        return Money<Currency>(-value.amount)
    }
}

extension Money {
    /// The product of a monetary amount and a scalar value.
    public static func * (lhs: Money<Currency>, rhs: Decimal) -> Money<Currency> {
        return Money<Currency>(lhs.amount * rhs)
    }

    /**
        The product of a monetary amount and a scalar value.

        - Important: Multiplying a monetary amount by a floating-point number
                     results in an amount rounded to the number of places
                     of the minor currency unit.
                     To produce a smaller fractional monetary amount,
                     multiply by a `Decimal` value instead.
     */
    public static func * (lhs: Money<Currency>, rhs: Double) -> Money<Currency> {
        return (lhs * Decimal(rhs)).rounded
    }

    /// The product of a monetary amount and a scalar value.
    public static func * (lhs: Money<Currency>, rhs: Int) -> Money<Currency> {
        return lhs * Decimal(rhs)
    }

    /// The product of a monetary amount and a scalar value.
    public static func * (lhs: Decimal, rhs: Money<Currency>) -> Money<Currency> {
        return rhs * lhs
    }

    /**
        The product of a monetary amount and a scalar value.

        - Important: Multiplying a monetary amount by a floating-point number
                     results in an amount rounded to the number of places
                     of the minor currency unit.
                     To produce a smaller fractional monetary amount,
                     multiply by a `Decimal` value instead.
     */
    public static func * (lhs: Double, rhs: Money<Currency>) -> Money<Currency> {
        return rhs * lhs
    }

    /// The product of a monetary amount and a scalar value.
    public static func * (lhs: Int, rhs: Money<Currency>) -> Money<Currency> {
        return rhs * lhs
    }

    /// Multiplies a monetary amount by a scalar value.
    public static func *= (lhs: inout Money<Currency>, rhs: Decimal) {
        lhs.amount *= rhs
    }

    /// Multiplies a monetary amount by a scalar value.
    /**
        Multiplies a monetary amount by a scalar value.

        - Important: Multiplying a monetary amount by a floating-point number
                     results in an amount rounded to the number of places
                     of the minor currency unit.
                     To produce a smaller fractional monetary amount,
                     multiply by a `Decimal` value instead.
     */
    public static func *= (lhs: inout Money<Currency>, rhs: Double) {
        lhs.amount = Money<Currency>(lhs.amount * Decimal(rhs)).rounded.amount
    }

    /// Multiplies a monetary amount by a scalar value.
    public static func *= (lhs: inout Money<Currency>, rhs: Int) {
        lhs.amount *= Decimal(rhs)
    }
}

// MARK: - CustomStringConvertible

extension Money: CustomStringConvertible {
    public var description: String {
        return "\(self.amount)"
    }
}

// MARK: - LosslessStringConvertible

extension Money: LosslessStringConvertible {
    public init?(_ description: String) {
        guard let amount = Decimal(string: description) else {
            return nil
        }

        self.init(amount)
    }
}

// MARK: - ExpressibleByIntegerLiteral

extension Money: ExpressibleByIntegerLiteral {
    public init(integerLiteral value: Int) {
        self.init(Decimal(integerLiteral: value))
    }
}

// MARK: - ExpressibleByFloatLiteral

extension Money: ExpressibleByFloatLiteral {
    /**
     Creates a new value from the given floating-point literal.
     
     - Important: Swift floating-point literals are currently initialized
                  using binary floating-point number type,
                  which cannot precisely express certain values.
                  As a workaround, monetary amounts initialized
                  from a floating-point literal are rounded
                  to the number of places of the minor currency unit.
                  To express a smaller fractional monetary amount,
                  initialize from a string literal or decimal value instead.
     - Bug: See https://bugs.swift.org/browse/SR-920
    */
    public init(floatLiteral value: Double) {
        self.init(Decimal(floatLiteral: value).rounded(for: Currency.self))
    }
}

// MARK: - ExpressibleByStringLiteral

extension Money: ExpressibleByStringLiteral {
    public init(unicodeScalarLiteral value: Unicode.Scalar) {
        self.init(stringLiteral: String(value))
    }

    public init(extendedGraphemeClusterLiteral value: Character) {
        self.init(stringLiteral: String(value))
    }

    public init(stringLiteral value: String) {
        self.init(value)!
    }
}

// MARK: - Codable

/**
 Coding keys for `Money` values.
 */
public enum MoneyCodingKeys: String, CodingKey {
    /// The coding key for the `amount` property.
    case amount

    /// The coding key for the `currencyCode` property
    case currencyCode
}

extension CodingUserInfoKey {
    /**
     The key for specifying custom decoding options for `Money` values.

     This user info key should be associated with
     an `MoneyDecodingOptions` object.

     - SeeAlso: `JSONDecoder.moneyDecodingOptions`
     */
    public static let moneyDecodingOptions = CodingUserInfoKey(rawValue: "com.flightschool.money.decoding-options")!

    /**
     The key for specifying custom encoding options for `Money` values.

     This user info key should be associated with
     an `MoneyEncodingOptions` object.

     - SeeAlso: `JSONDecoder.moneyEncodingOptions`
     */
    public static let moneyEncodingOptions = CodingUserInfoKey(rawValue: "com.flightschool.money.encoding-options")!
}

/**
 Custom decoding options for `Money` values.

 Configure the decoding behavior either
 by using the `JSONDecoder.moneyDecodingOptions` property
 or by setting the `CodingUserInfoKey.moneyDecodingOptions` key
 in the decoder's `userInfo` property.
 */
public struct MoneyDecodingOptions: OptionSet {
    public let rawValue: Int

    public init(rawValue: Int) {
        self.rawValue = rawValue
    }

    /**
     Throws a `DecodingError` when attempting to decode
     a `Money` value from a string or number value.

     By default, `Money` values are decoded from single values
     using the associated `Currency` type.
     */
    public static let requireExplicitCurrency = MoneyDecodingOptions(rawValue: 1 << 0)

    /**
     Throws an error when attempting to decode
     `amount` from a floating-point number.

     - Important: Foundation decoders currently decode number values
                  using a binary floating-point number type,
                  which cannot precisely express certain values.
                  Specify this option to require monetary amounts
                  to be decoded precisely from string representations.
     - Bug: See https://bugs.swift.org/browse/SR-7054.
     */
    public static let requireStringAmount = MoneyDecodingOptions(rawValue: 1 << 1)

    /**
     Rounds `amount` to the number of places of the minor currency unit
     when decoding from a floating-point number.
     */
    public static let roundFloatingPointAmount = MoneyDecodingOptions(rawValue: 1 << 2)
}

/**
 Custom encoding options for `Money` values.

 Configure the encoding behavior either
 by using the `JSONDecoder.moneyEncodingOptions` property
 or by setting the `CodingUserInfoKey.moneyEncodingOptions` key
 in the encoder's `userInfo` property.
 */
public struct MoneyEncodingOptions: OptionSet {
    public let rawValue: Int

    public init(rawValue: Int) {
        self.rawValue = rawValue
    }

    /**
     Encodes the `Money` value as a single value,
     without specifying a currency.
     */
    public static let omitCurrency = MoneyEncodingOptions(rawValue: 1 << 0)


    /**
     Encodes the string representation of `amount`
     instead of the built-in `Decimal` encoding.
    */
    public static let encodeAmountAsString = MoneyEncodingOptions(rawValue: 1 << 1)
}

extension Money: Codable {
    public init(from decoder: Decoder) throws {
        let options = decoder.userInfo[.moneyDecodingOptions] as? MoneyDecodingOptions ?? []

        if let keyedContainer = try? decoder.container(keyedBy: MoneyCodingKeys.self) {
            let currencyCode = try keyedContainer.decode(String.self, forKey: .currencyCode)
            guard currencyCode == Currency.code else {
                let context = DecodingError.Context(codingPath: keyedContainer.codingPath, debugDescription: "Currency mismatch: expected \(Currency.code), got \(currencyCode)")
                throw DecodingError.typeMismatch(Money<Currency>.self, context)
            }

            var amount: Decimal?
            if let string = try? keyedContainer.decode(String.self, forKey: .amount) {
                amount = Decimal(string: string)
            } else if !options.contains(.requireStringAmount) {
                amount = try keyedContainer.decode(Decimal.self, forKey: .amount)
                if options.contains(.roundFloatingPointAmount) {
                    amount = amount?.rounded(for: Currency.self)
                }
            }

            if let amount = amount {
                self.amount = amount
            } else {
                throw DecodingError.dataCorruptedError(forKey: .amount, in: keyedContainer, debugDescription: "Couldn't decode Decimal value for amount")
            }
        } else if !options.contains(.requireExplicitCurrency),
            let singleValueContainer = try? decoder.singleValueContainer()
        {
            var amount: Decimal?
            if let string = try? singleValueContainer.decode(String.self) {
                amount = Decimal(string: string)
            } else if !options.contains(.requireStringAmount) {
                amount = try singleValueContainer.decode(Decimal.self)
                if options.contains(.roundFloatingPointAmount) {
                    amount = amount?.rounded(for: Currency.self)
                }
            }

            if let amount = amount {
                self.amount = amount
            } else {
                throw DecodingError.dataCorruptedError(in: singleValueContainer, debugDescription: "Couldn't decode Decimal value for amount")
            }
        } else {
            let context = DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Couldn't decode Money value")
            throw DecodingError.dataCorrupted(context)
        }
    }

    public func encode(to encoder: Encoder) throws {
        let options = encoder.userInfo[.moneyEncodingOptions] as? MoneyEncodingOptions ?? []

        if options.contains(.omitCurrency) {
            var singleValueContainer = encoder.singleValueContainer()
            if options.contains(.encodeAmountAsString) {
                try singleValueContainer.encode(self.amount.description)
            } else {
                try singleValueContainer.encode(self.amount)
            }
        } else {
            var keyedContainer = encoder.container(keyedBy: MoneyCodingKeys.self)
            try keyedContainer.encode(Currency.code, forKey: .currencyCode)
            if options.contains(.encodeAmountAsString) {
                try keyedContainer.encode(self.amount.description, forKey: .amount)
            } else {
                try keyedContainer.encode(self.amount, forKey: .amount)
            }
        }
    }
}

// MARK: -

fileprivate extension Decimal {
    func rounded(for currency: CurrencyType.Type) -> Decimal {
        var approximate = self
        var rounded = Decimal()
        NSDecimalRound(&rounded, &approximate, currency.minorUnit, .bankers)

        return rounded
    }
}


Money.playground/Contents.swift
import Foundation
import Money

let amount = Decimal(12)
let monetaryAmount = Money<USD>(amount)

12 as Money<USD>
12.00 as Money<USD>
"12.00" as Money<USD>

let preciseAmount: Money<USD> = "123.4567"
let roundedAmount: Money<USD> = 123.4567 // rounded to minor unit

preciseAmount.amount // 123.4567
roundedAmount.amount // 123.46

let dollarAmount: Money<USD> = 123.45
let euroAmount: Money<EUR> = 4567.89
let yenAmount: Money<JPY> = 8100
let canadianDollarAmount: Money<CAD> = 234

// dollarAmount == euroAmount // Error: Binary operator '==' cannot be applied

let prices: [Money<USD>] = [2.19, 5.39, 20.99, 2.99, 1.99, 1.99, 0.99]
let subtotal = prices.reduce(0.00, +)
let tax = 0.08 * subtotal
let total = subtotal + tax

let formatter = NumberFormatter()
formatter.numberStyle = .currency
formatter.locale = Locale(identifier: "fr-FR")
formatter.currencyCode = total.currency.code
formatter.string(for: total.amount)
